homeassistant/components/myuplink/coordinator.py

```diff
@@ -0,0 +1,95 @@
+"""Coordinator for myUplink."""
+from datetime import datetime, timedelta
+import logging
+
+import async_timeout
+from myuplink.api import MyUplinkAPI
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import (
+    MU_DATAGROUP_DEVICES,
+    MU_DATAGROUP_POINTS,
+    MU_DATAGROUP_SYSTEMS,
+    MU_DATATIME,
+    MU_DEVICE_CONNECTIONSTATE,
+    MU_DEVICE_FIRMWARE_CURRENT,
+    MU_DEVICE_FIRMWARE_DESIRED,
+    MU_DEVICE_PRODUCTNAME,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class MyUplinkDataCoordinator(DataUpdateCoordinator):
+    """Coordinator for myUplink data."""
+
+    def __init__(self, hass: HomeAssistant, mu_api: MyUplinkAPI) -> None:
+        """Initialize myUplink coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name="myuplink",
+            update_interval=timedelta(seconds=60),
+        )
+        self.mu_api = mu_api
+
+    async def _async_update_data(self):
+        """Fetch data from the myUplink API."""
+        async with async_timeout.timeout(10):
+
+            _LOGGER.debug("Coordinator preparing updating")
+
+            data = {}
+
+            # Get systems
+            mu_systems = await self.mu_api.async_get_systems()
+            data[MU_DATAGROUP_SYSTEMS] = mu_systems
+
+            # Get device info
+            devices = {}
+            device_ids = await self.get_system_device_ids()
+            for device_id in device_ids:
+                api_device_info = await self.mu_api.async_get_device(device_id)
+                device_info = {}
+                device_info[
+                    MU_DEVICE_FIRMWARE_CURRENT
+                ] = api_device_info.firmwareCurrent
+                device_info[
+                    MU_DEVICE_FIRMWARE_DESIRED
+                ] = api_device_info.firmwareDesired
+                device_info[MU_DEVICE_CONNECTIONSTATE] = api_device_info.connectionState
+                device_info[MU_DEVICE_PRODUCTNAME] = api_device_info.productName
+                devices[device_id] = device_info
+
+            # Get device points (data)
+            points = {}
```

Comment:
I don't think you have to, but it would be good to type all these dictionaries you are making

Reply:
Can you give an example? It's dynamic dictionaries.

Reply:
point_info: dict[str, DevicePoint]
points: dict[str, dict[str, DevicePoint]]

Again II don't think it is needed, but the more typing the better imo, esp if you ever want to enable strict typing.